contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -1,452 +1,584 @@
         external
-        authP(MANAGE_SIGNING_KEYS, arr(_operator_id))
+        authP(SET_NODE_OPERATOR_LIMIT_ROLE, arr(uint256(_nodeOperatorId), uint256(_vettedSigningKeysCount)))
+        onlyExistedNodeOperator(_nodeOperatorId)
     {
-        _removeSigningKey(_operator_id, _index);
+        NodeOperator storage nodeOperator = _nodeOperators[_nodeOperatorId];
+
+        require(nodeOperator.active, "NODE_OPERATOR_DEACTIVATED");
+
+        uint64 totalSigningKeysCount = nodeOperator.totalSigningKeysCount;
+        uint64 depositedSigningKeysCount = nodeOperator.depositedSigningKeysCount;
+        uint64 vettedSigningKeysCountBefore = nodeOperator.vettedSigningKeysCount;
+
+        uint64 vettedSigningKeysCountAfter = Math64.min(
+            totalSigningKeysCount,
+            Math64.max(_vettedSigningKeysCount, depositedSigningKeysCount)
+        );
+
+        if (vettedSigningKeysCountAfter == vettedSigningKeysCountBefore) {
+            return;
+        }
+
+        nodeOperator.vettedSigningKeysCount = vettedSigningKeysCountAfter;
+
+        SigningKeysStats.State memory totalSigningKeysStats = _getTotalSigningKeysStats();
+        if (vettedSigningKeysCountAfter > vettedSigningKeysCountBefore) {
+            totalSigningKeysStats.increaseVettedSigningKeysCount(vettedSigningKeysCountAfter - vettedSigningKeysCountBefore);
+        } else {
+            totalSigningKeysStats.decreaseVettedSigningKeysCount(vettedSigningKeysCountBefore - vettedSigningKeysCountAfter);
+        }
+        _setTotalSigningKeysStats(totalSigningKeysStats);
+
+        emit VettedSigningKeysCountChanged(_nodeOperatorId, vettedSigningKeysCountAfter);
+        _increaseValidatorsKeysNonce();
     }
 
-    /**
-      * @notice Removes an #`_amount` of validator signing keys starting from #`_index` of operator #`_id` usable keys. Executed on behalf of DAO.
-      * @param _operator_id Node Operator id
-      * @param _index Index of the key, starting with 0
-      * @param _amount Number of keys to remove
-      */
-    function removeSigningKeys(uint256 _operator_id, uint256 _index, uint256 _amount)
+    /// @notice Updates the number of the validators in the EXITED state for node operator with given id
+    /// @param _nodeOperatorId Id of the node operator
+    /// @param _exitedValidatorsKeysCount New number of EXITED validators of the node operator
+    function updateExitedValidatorsKeysCount(uint256 _nodeOperatorId, uint256 _exitedValidatorsKeysCount)
         external
-        authP(MANAGE_SIGNING_KEYS, arr(_operator_id))
+        onlyExistedNodeOperator(_nodeOperatorId)
+        auth(UPDATE_EXITED_VALIDATORS_KEYS_COUNT_ROLE)
     {
-        // removing from the last index to the highest one, so we won't get outside the array
-        for (uint256 i = _index.add(_amount); i > _index; --i) {
-            _removeSigningKey(_operator_id, i - 1);
+        require(_exitedValidatorsKeysCount <= UINT64_MAX, "EXITED_VALIDATORS_COUNT_TOO_LARGE");
```

Comment:
extra check because of require(_exitedValidatorsKeysCount <= depositedSigningKeysCount, "INVALID_EXITED_VALIDATORS_COUNT");